packages/sitecore-jss/src/graphql-request-client.ts

```diff
@@ -93,23 +107,48 @@ export class GraphQLRequestClient implements GraphQLClient {
         query,
         variables,
       });
+      let retriesLeft = this.retries || 1;
 
-      const fetchWithOptionalTimeout = [this.client.request(query, variables)];
-      if (this.timeout) {
-        this.abortTimeout = new TimeoutPromise(this.timeout);
-        fetchWithOptionalTimeout.push(this.abortTimeout.start);
-      }
-      Promise.race(fetchWithOptionalTimeout).then(
-        (data: T) => {
-          this.abortTimeout?.clear();
-          this.debug('response in %dms: %o', Date.now() - startTimestamp, data);
-          resolve(data);
-        },
-        (error: ClientError) => {
-          this.abortTimeout?.clear();
-          this.debug('response error: %o', error.response || error.message || error);
-          reject(error);
+      const retryer = async (): Promise<T> => {
+        const startTimestamp = Date.now();
+        retriesLeft--;
+        const fetchWithOptionalTimeout = [this.client.request(query, variables)];
+        if (this.timeout) {
+          this.abortTimeout = new TimeoutPromise(this.timeout);
+          fetchWithOptionalTimeout.push(this.abortTimeout.start);
         }
+        return Promise.race(fetchWithOptionalTimeout).then(
+          (data: T) => {
+            this.abortTimeout?.clear();
+            this.debug('response in %dms: %o', Date.now() - startTimestamp, data);
+            return Promise.resolve(data);
+          },
+          (error: ClientError) => {
+            this.abortTimeout?.clear();
+            this.debug('response error: %o', error.response || error.message || error);
+            if (error.response?.status === 429 && retriesLeft > 0) {
+              const rawHeaders = (error as ClientError)?.response?.headers;
+              const delaySeconds =
+                rawHeaders && rawHeaders.get('Retry-After')
+                  ? Number.parseInt(rawHeaders.get('Retry-After'), 10)
+                  : 1;
+              this.debug(
+                'Endpoint responded with 429. Retrying in %ds. Retries left: %d',
```

Comment:
Do we need to specify what "429" code means? The same case as I mentioned above, a customer will need to search the definition.
```suggestion
                'You exceeded a rate limit of requests. Retrying in %ds. Retries left: %d',
```

Reply:
Used slightly different wording here - but it should convey an error and the cause clearly for it now